UI: Streaming Settings Validation On Stream Start & User Help

JohannMG opened the pull request:
<!--- Please fill out the following template, which will help other contributors review your Pull Request. -->

<!--- Make sure you’ve read the contribution guidelines here: https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst -->

### Description
<!--- Describe your changes in detail. -->
<!--- If this change includes UI elements, please include screenshots. -->

The first commit PR moves UI verification to another class to be reusable.
The second commit adds checks to "Start Streaming" user action prompting users to fill out streaming URL, service, and stream keys if they're missing. 


![Screen Shot 2019-11-26 at 5 28 14 PM](https://user-images.githubusercontent.com/1580980/69685903-de4b1480-1072-11ea-8903-24350f800631.png)

![Screen Shot 2019-11-26 at 5 19 28 PM](https://user-images.githubusercontent.com/1580980/69685904-dee3ab00-1072-11ea-94b2-fcf74d02694e.png)



### Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open Mantis issue, or implements feature request -->
<!--- from the Ideas page, please link to the issue here. -->

When we use the server error to respond to user's failing to start a
stream it is slow to return an error and unclear to the user what went
wrong. 

In the case of preset services (Youtube, Facebook, Twitch) we check there is a Stream Key before
attempting to start a stream.

In the case of "custom" we only verify there is a URL since for some
services that's all that is required or the user may use user/password
authentication.

In both cases the short explanation and Settings link button will guide users to solve the issue more smoothly. 

### How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment (hardware, OS version, etc.),-->
<!--- and the tests you ran, including how it may affect other areas of code. -->

MacbookPro. MacOSX 10.14.6. 
- Tested custom stream with no URL in settings prompts the user.
- Tested custom stream with URL in setting does not prompt the user.
- Tested preset service with no Stream Key prompts the user.
- Tested preset service with any Stream Key does not prompt the user. 
- The "do you want to start" stream button still works. 
- Does not affect stopping the stream. 
- Start/Stop Streaming button is reset when streaming is aborted. 

### Types of changes
<!--- What types of changes does your PR introduce? Uncomment all that apply -->
<!--- - Bug fix (non-breaking change which fixes an issue) -->
<!--- - New feature (non-breaking change which adds functionality) -->
<!--- - Performance enhancement (non-breaking change which improves efficiency) -->
<!--- - Code cleanup (non-breaking change which makes code smaller or more readable) -->
<!--- - Breaking change (fix or feature that would cause existing functionality to change) -->
<!--- - Documentation (a change to documentation pages) -->

 - Bug fix (non-breaking change which fixes an issue)
 - New feature (non-breaking change which adds functionality)
 - UI Improvement 

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.

JohannMG commented:
Want to do in the future and out of scope here for now: 
 - Add in-settings highlighting of empty require fields
 - Have settings button that will open to a specific tab.

WizardCM commented:
Only one functional strangeness: if you choose "Settings", the "Start Streaming" button remains highlighted until you close the Settings dialog. Otherwise, looks good! Not 100% sold on the wording, but that can always be tweaked.

Tested on Windows 10, here's how they look (ignore the large size, that's just High DPI):

![](http://scr.wzd.li/scr/2019-11-27_10-57-08.png)

![](http://scr.wzd.li/scr/2019-11-27_10-57-49.png)

JohannMG commented:
Hate how the Mac buttons are ordered differently...
Let see if I can fix that button and also talking to a UX designer right now on better wording may update this later today/tomorrow.

WizardCM commented:
Yeah, ordering is strange on Linux too in my experience. I do wonder if it'd be possible to put the Settings button on the bottom left corner, kind of like this:

![](https://miro.medium.com/max/1032/1*ts3nrHe_b9xaN46pE7-EWg.png)

JohannMG commented:
> I do wonder if it'd be possible to put the Settings button on the bottom left corner, kind of like this. 

@WizardCM Ha. that's what my designer and I were exactly aiming for but there's a few caveats in Qt around this apparently. Except to east the user path to stream, fixing settings would be the primary action.  

I can't find a smooth way to do this in Qt...

WizardCM commented:
> @WizardCM Ha. that's what my designer and I were exactly aiming for but there's a few caveats in Qt around this apparently. Except to east the user path to stream, fixing settings would be the primary action.
> 
> I can't find a smooth way to do this in Qt...

Common enough in Qt. Let's look at it in a different way then - would it make sense to remove the "Continue" button altogether? A required field is a required field, after all. Or is the goal to handle edge cases and hacks that third parties may have put together in their own plugins/outputs?

JohannMG commented:
Yes, the goal to handle edge cases and hacks. I don't want to accidentally completely block someone's functionality.

JohannMG commented:
Do you think the continue is important? **I would be happy to remove it.** 
Just worried there may be edge cases.

JohannMG commented:
Updated: 

 - Change copy, you're right used some polish. Talked to design. 
 - Moved Cancel to the left, making Settings more clearly the path to fixing this issue
 - Removed the "continue" path because I made the stream check more loose for "custom" so it shouldn't be a problem. 
- Updates first post. 


![Screen Shot 2019-11-26 at 5 28 14 PM](https://user-images.githubusercontent.com/1580980/69685903-de4b1480-1072-11ea-8903-24350f800631.png)

![Screen Shot 2019-11-26 at 5 19 28 PM](https://user-images.githubusercontent.com/1580980/69685904-dee3ab00-1072-11ea-94b2-fcf74d02694e.png)


>  if you choose "Settings", the "Start Streaming" button remains highlighted until you close the Settings dialog.

So double checked I am setting the button back to unchecked but the main UI isn't refreshed in the background.

tt2468 commented:
Would it be possible to add a "Do not show again" option for stream key missing? You can stream to for example nginx rtmp without using a stream name/key. At the very least make it possible to override the error.

JohannMG commented:
@tt2468 

From description: 
> In the case of "custom" we only verify there is a URL since for some services that's all that is required or the user may use user/password authentication.

This case is already included: for custom URL addresses it only verifies there's a URL to send to assuming you can put all auth into in the custom URL. You will not see the popup. (I have used a similar setup myself)

In code it's here were "isCustomUrlService" overrides the check for a stream key: https://github.com/obsproject/obs-studio/pull/2203/files#diff-2f02d351252e2849eb4e7b684f0777adR72

notr1ch commented:
Having Cancel be the primary button on the left is very jarring to me as a Windows user. Apparently Apple always put Cancel first in dialogs which is naturally the opposite of how Windows does it. Is there a way to have Qt take this into account so it remains consistent with the OS?

jp9000 commented:
Great PR!  I'm very happy to see nice user experience fixups.

I agree with r1ch, and we might want to check to see how Linux normally handles that too (I think Linux is the same as Windows in this regard).  Basically it should probably just check for an __APPLE__ macro definition and then swap it if it's not apple, I think.

JohannMG commented:
It's aligned by button role in a QMessageBox, not manually: can someone pull this branch and check if it's left aligned on Windows/Linux b/c QMessageBox should adjust per platform on it's own (and I don't have VMs for either ready at the moment.

jp9000 commented:
Oh woops, I'll give it a check, I didn't know that.

jp9000 commented:
Here you go:
![obs64_2019-11-27_18-48-58](https://user-images.githubusercontent.com/2084112/69773064-a01f2500-1146-11ea-9ed0-cb872b1c7857.png)

JohannMG commented:
Jim, what order would you want this in to approve? I'll revisit this after thanksgiving madness here :)

jp9000 commented:
I think if it's windows and linux, I think cancel should always be on the right, if it's mac then the cancel button should be on the left.  Checking for the __APPLE__ macro should be sufficient

jp9000 commented:
Oh and no problem, have a great holiday!

jp9000 commented:
Just wanted to check back with you.  If you're busy, would you mind if I edited it to swap the buttons depending on operating system and merged it?

JohannMG commented:
Go ahead! I'm out of office for few days longer so feel free ✅

On Tue, Dec 10, 2019 at 17:14 Jim <notifications@github.com> wrote:

> Just wanted to check back with you. If you're busy, would you mind if I
> edited it to swap the buttons depending on operating system and merged it?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/obsproject/obs-studio/pull/2203?email_source=notifications&email_token=AAMB7NGIYAQAKX2N7ALLFTDQYAPDZA5CNFSM4JR7IG3KYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEGRJGFQ#issuecomment-564302614>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAMB7NFOVL4TBHVT3VLHINLQYAPDZANCNFSM4JR7IG3A>
> .
>

JohannMG commented:
❤️ NICE
